File: app/jobs/geocoding_job.rb
class GeocodingJob < ApplicationJob
  queue_as :default
  retry_on StandardError, wait: ->(executions) { (executions ** 4) + rand(30) }, attempts: 3

  # Don't retry if contact was deleted
  discard_on ActiveRecord::RecordNotFound do |job, exception|
    Rails.logger.error("[GeocodingJob] Contact not found: #{exception.message}")
  end

  def perform(contact_id)
    contact = Contact.find(contact_id)
    credentials = TwilioCredential.current

    return unless credentials&.enable_geocoding

    service = GeocodingService.new(contact)
    result = service.geocode!

    if result[:success]
      Rails.logger.info "Successfully geocoded contact #{contact_id}: #{result[:latitude]}, #{result[:longitude]}"
    else
      Rails.logger.warn "Geocoding failed for contact #{contact_id}: #{result[:error]}"
    end

    result
  rescue StandardError => e
    Rails.logger.error "Geocoding job failed for contact #{contact_id}: #{e.message}"
    raise
  end
end

File: laravel-app/app/Jobs/GeocodingJob.php
<?php

namespace App\Jobs;

use App\Models\Contact;
use App\Models\TwilioCredential;
use App\Services\GeocodingService;
use Illuminate\Bus\Queueable;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Foundation\Bus\Dispatchable;
use Illuminate\Queue\InteractsWithQueue;
use Illuminate\Queue\SerializesModels;
use Illuminate\Support\Facades\Log;
use Exception;

class GeocodingJob implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;

    public $contactId;

    /**
     * The number of times the job may be attempted.
     *
     * @var int
     */
    public $tries = 3;

    /**
     * The number of seconds to wait before retrying.
     *
     * @var int
     */
    public $backoff = [10, 30, 90];

    /**
     * The maximum number of seconds the job can run before timing out.
     *
     * @var int
     */
    public $timeout = 60;

    /**
     * Create a new job instance.
     *
     * @param int $contactId
     */
    public function __construct(int $contactId)
    {
        $this->contactId = $contactId;
        $this->onQueue('default');
    }

    /**
     * Execute the job.
     *
     * @return array
     */
    public function handle()
    {
        try {
            $contact = Contact::findOrFail($this->contactId);
            $credentials = TwilioCredential::current();

            if (!$credentials || !$credentials->enable_geocoding) {
                return ['success' => false, 'error' => 'Geocoding disabled'];
            }

            $service = new GeocodingService($contact);
            $result = $service->geocode();

            if ($result['success']) {
                Log::info("Successfully geocoded contact {$this->contactId}: {$result['latitude']}, {$result['longitude']}");
            } else {
                Log::warning("Geocoding failed for contact {$this->contactId}: {$result['error']}");
            }

            return $result;

        } catch (Exception $e) {
            Log::error("Geocoding job failed for contact {$this->contactId}: {$e->getMessage()}");
            throw $e;
        }
    }

    /**
     * Handle a job failure.
     *
     * @param  \Throwable  $exception
     * @return void
     */
    public function failed(\Throwable $exception)
    {
        Log::error("[GeocodingJob] Failed for contact {$this->contactId}: {$exception->getMessage()}");
    }
}
